### PR TITLE
trigger unsized coercions keyed on Sized bounds

### LINKED ISSUE #49593: Box::new(!) doesn't coerce to `Box<Error>` successfully
[Discovered](https://github.com/rust-lang/rust/pull/49039#issuecomment-376447044) in #49039 the recent stabilization of `!` unfortunately has some hazards related to coercion on the libs side of things. We discussed this at the work week and decided that we can't land PRs like #49039 which would *also* mean that the libs team wouldn't be able to actually recommend using `!` as an error type (bummer!)

This is intended to be a tracking issue for sorting this out. I'm tagging it as a regression to mainly keep track of it, but it's not strictly speaking a regression.

cc @rust-lang/libs 
cc @rust-lang/lang

### PR DESCRIPTION
**Pull request by arielb1**
This PR causes unsized coercions to not be disabled by `$0: Unsize<dyn
Object>` coercion obligations when we have an `$0: Sized` obligation somewhere.

This should be mostly backwards-compatible, because in these cases not doing the unsize coercion should have caused the `$0: Sized` obligation to fail.

Note that `X: Unsize<dyn Object>` obligations can't fail *as obligations* if `X: Sized` holds, so this still maintains some version of monotonicity (I think that an unsized coercion can't be converted to no coercion by unifying type variables).

Fixes #49593 (unblocking never_type).

r? @eddyb 
cc @nikomatsakis

### CODE REVIEW

File: src/librustc_typeck/check/mod.rs [hunk 1]
```diff
@@ -2717,6 +2719,97 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
...
+        let self_ty = self.shallow_resolve(trait_ref.self_ty());
+        debug!(
+            "self_type_matches_expected_vid(trait_ref={:?}, self_ty={:?}, expected_vid={:?})",
+            trait_ref, self_ty, expected_vid
+        );
+        match self_ty.sty {
+            ty::Infer(ty::TyVar(v)) => {
+                let root_vid = self.root_var(v);
+                debug!("self_type_matches_expected_vid - root_vid={:?}", root_vid);
+                if root_vid == expected_vid {
```
Comment (Centril) [inline src/librustc_typeck/check/mod.rs]: `if` expr seems redundant?

File: src/test/ui/coercion/coerce-issue-49593-box-never.rs [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+#![allow(unreachable_code)]
+
+use std::error::Error;
+use std::char::ParseCharError; /* some Error */
+
+fn raw_ptr_box<T>(t: T) -> *mut T {
+    panic!()
+}
+
+fn foo(x: !) -> Box<Error> {
```
Comment (Centril) [inline src/test/ui/coercion/coerce-issue-49593-box-never.rs]:
```suggestion
fn foo(x: !) -> Box<dyn Error> {
```

File: src/test/ui/coercion/coerce-issue-49593-box-never.rs [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    /* *mut $0 is coerced to *mut Error here */ Box::<_ /* ! */>::new(x)
+}
+
+fn foo_raw_ptr(x: !) -> *mut Error {
```
Comment (Centril) [inline src/test/ui/coercion/coerce-issue-49593-box-never.rs]:
```suggestion
fn foo_raw_ptr(x: !) -> *mut dyn Error {
```

File: src/test/ui/coercion/coerce-issue-49593-box-never.rs [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    /* *mut $0 is coerced to *mut Error here */ raw_ptr_box::<_ /* ! */>(x)
+}
+
+fn no_coercion(d: *mut Error) -> *mut Error {
```
Comment (Centril) [inline src/test/ui/coercion/coerce-issue-49593-box-never.rs]:
```suggestion
fn no_coercion(d: *mut dyn Error) -> *mut dyn Error {
```

File: src/test/ui/coercion/coerce-issue-49593-box-never.rs [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+struct T;
+impl Xyz for S {}
+impl Xyz for T {}
+
+fn foo_no_never() {
+    let mut x /* : Box<S> */ = None;
+    let mut first_iter = false;
+    loop {
+        if !first_iter {
+            let y: Box<Xyz>
```
Comment (Centril) [inline src/test/ui/coercion/coerce-issue-49593-box-never.rs]:
```suggestion
            let y: Box<dyn Xyz>
```

File: src/test/ui/coercion/coerce-issue-49593-box-never.rs [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+fn raw_ptr_box<T>(t: T) -> *mut T {
+    panic!()
+}
+
+fn foo(x: !) -> Box<Error> {
+    /* *mut $0 is coerced to *mut Error here */ Box::<_ /* ! */>::new(x)
+}
+
+fn foo_raw_ptr(x: !) -> *mut Error {
+    /* *mut $0 is coerced to *mut Error here */ raw_ptr_box::<_ /* ! */>(x)
```
Comment (Centril) [inline src/test/ui/coercion/coerce-issue-49593-box-never.rs]:
```suggestion
    /* *mut $0 is coerced to *mut dyn Error here */ raw_ptr_box::<_ /* ! */>(x)
```

File: src/test/ui/coercion/coerce-issue-49593-box-never.rs [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+use std::error::Error;
+use std::char::ParseCharError; /* some Error */
+
+fn raw_ptr_box<T>(t: T) -> *mut T {
+    panic!()
+}
+
+fn foo(x: !) -> Box<Error> {
+    /* *mut $0 is coerced to *mut Error here */ Box::<_ /* ! */>::new(x)
```
Comment (Centril) [inline src/test/ui/coercion/coerce-issue-49593-box-never.rs]: Comment seems strange... there's no *mut T or *mut Error here?

File: src/test/ui/coercion/coerce-issue-49593-box-never.rs [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+}
+
+trait Xyz {}
+struct S;
+struct T;
+impl Xyz for S {}
+impl Xyz for T {}
+
+fn foo_no_never() {
+    let mut x /* : Box<S> */ = None;
```
Comment (Centril) [inline src/test/ui/coercion/coerce-issue-49593-box-never.rs]:
```suggestion
    let mut x /* : Option<Box<S>> */ = None;
```

**Comment by Centril** [comment]
> This should be mostly backwards-compatible, because in these cases not doing the unsize coercion should have caused the $0: Sized obligation to fail.

We should test that with a crater run?

File: src/librustc_typeck/check/mod.rs (hunk 1 above)
Comment (arielb1) [reply src/librustc_typeck/check/mod.rs]: Why?

Comment (arielb1) [reply src/librustc_typeck/check/mod.rs]: Ah ok there was more code there.

**Comment by arielb1** [comment]
> What did you try? You might've had to use `+ Captures<'tcx>` etc.

I first tried to have the function return `impl Iterator<Item=ty::PolyTraitRef<'tcx>> + 'b`. That gave me the following error:

```
error[E0700]: hidden type for `impl Trait` captures lifetime that does not appear in bounds  
    --> src/librustc_typeck/check/mod.rs:2788:12
     |
2788 |         -> impl Iterator<Item=ty::PolyTraitRef<'tcx>> + 'b
     |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
     |
note: hidden type `std::iter::FilterMap<std::vec::IntoIter<rustc::traits::Obligation<'tcx, rustc::ty::Predicate<'tcx>>>, check::ObligationMapper<'b, 'gcx, 'tcx>>` captures the lifetime 'gcx as defined on the impl at 2786:10

```

I then tried to use the `impl Iterator<Item=ty::PolyTraitRef<'tcx>> + 'a + 'b + 'gcx + 'tcx` syntax, which gave me:
```
error: unsatisfied lifetime constraints==========================>   ] 106/111: rustc_ty...  
    --> src/librustc_typeck/check/mod.rs:2788:12
     |
2786 | impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
     |          ----  ---- lifetime `'tcx` defined here
     |          |
     |          lifetime `'gcx` defined here
2787 |     fn obligations_for_self_ty<'b>(&'b self, self_ty: ty::TyVid)
2788 |         -> impl Iterator<Item=ty::PolyTraitRef<'tcx>> + 'a + 'b + 'gcx + 'tcx
     |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ requires that `'tcx` must outlive `'gcx`

error: unsatisfied lifetime constraints
    --> src/librustc_typeck/check/mod.rs:2795:9
     |
2786 |   impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
     |        --  ---- lifetime `'gcx` defined here
     |        |
     |        lifetime `'a` defined here
...
2795 | /         self.fulfillment_cx
2796 | |             .borrow()
2797 | |             .pending_obligations()
2798 | |             .into_iter()
2799 | |             .filter_map(ObligationMapper(self, ty_var_root))
     | |____________________________________________________________^ returning this value requires that `'a` must outlive `'gcx`

error: unsatisfied lifetime constraints
    --> src/librustc_typeck/check/mod.rs:2795:9
     |
2786 |   impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
     |        -- lifetime `'a` defined here
2787 |       fn obligations_for_self_ty<'b>(&'b self, self_ty: ty::TyVid)
     |                                  -- lifetime `'b` defined here
...
2795 | /         self.fulfillment_cx
2796 | |             .borrow()
2797 | |             .pending_obligations()
2798 | |             .into_iter()
2799 | |             .filter_map(ObligationMapper(self, ty_var_root))
     | |____________________________________________________________^ returning this value requires that `'b` must outlive `'a`
```

Of course, my original design had closures instead of a struct implementing `FnMut`, but when I didn't have issues with `move` they gave a similar set of errors.

**Comment by arielb1** [comment]
So does this require a lang-team FCP?

**Comment by Centril** [comment]
@arielb1 I think if not FCP we should at least discuss it (and so I've nominated it) since this has user observable type system changes.

**Comment by nikomatsakis** [comment]
We had a discussion in the @rust-lang/lang meeting and concluded that we should go ahead with this change, at least in some form. Although we are now taking more things into consideration for coercions (namely, pending coercions) and that's a bit risky, it's nice to fix a real problem, and this doesn't seem like a "fundamental complexity change" in the coercion rules. They've always taken into account "information gained from type-checking thus far" and now we are just expanding that to the "complete set of information" (i.e. not just equality constraints, but arbitrary obligations).

I'd still personally like to spend some time tomorrow to think over some examples and try to be careful about precisely which obligations we consider, but the lang team is 👍 on the general direction thus far. =)

File: src/librustc_typeck/check/closure.rs [hunk 9]
```diff
@@ -235,13 +236,14 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
...
                     // the complete signature.
                     let trait_ref = proj_predicate.to_poly_trait_ref(self.tcx);
-                    self.self_type_matches_expected_vid(trait_ref, expected_vid)
-                        .and_then(|_| {
-                            self.deduce_sig_from_projection(
-                                Some(obligation.cause.span),
-                                proj_predicate
-                            )
-                        })
+                    Some(()).filter(|()| {
```
Comment (nikomatsakis) [inline src/librustc_typeck/check/closure.rs]: What is the purpose of this construct? I suspect there's a clearer one out there =)

**Review by nikomatsakis — APPROVED** [review]
OK, I read over the changes, which look good.

My only real hesitation is whether we want to try to be "intensional" about the set of obligations we consider for this role -- i.e., to trace the set of obligations that arise from the surrounding context in the same way we trace the 'expected type'. I believe that the existing closure checking code, which also uses obligations, was sort of aiming for this sort of limitation by tying itself to the case where (a) you have an expected type of an inference variable and (b) you have obligations directly referencing that variable. Pretty hacky, though.

I want to experiment a bit with this branch -- which for example checks for unifications of the expected type inference variable -- to see which things work and which things don't, and if there's much real change. I'm building a local copy for this purpose.

**Comment by arielb1** [comment]
> My only real hesitation is whether we want to try to be "intensional" about the set of obligations we consider for this role -- i.e., to trace the set of obligations that arise from the surrounding context in the same way we trace the 'expected type'. I believe that the existing closure checking code, which also uses obligations, was sort of aiming for this sort of limitation by tying itself to the case where (a) you have an expected type of an inference variable and (b) you have obligations directly referencing that variable. Pretty hacky, though.

Note that the obligations here are on the "found type" (i.e., the result of `Box::new`, which will be `Box<$0> where $0: Sized`), not on the expected type.

> (b) you have obligations directly referencing that variable

I added the `ty_var_root` normalization to the closure code too. Which type variable is selected as a root when unifying is sufficiently erratic that I don't think relying on it will do any good. If there's an edge case where you do want to rely on it, maybe a test/comment would be better?

**Comment by cramertj** [comment]
@nikomatsakis 
> I want to experiment a bit with this branch -- which for example checks for unifications of the expected type inference variable -- to see which things work and which things don't, and if there's much real change. I'm building a local copy for this purpose.

Ping on this-- did you have a chance to experiment? I'd love to see this land, since it unblocks both `never_type` and `TryFrom`/`TryInto` traits, AFAICT.

**Comment by nikomatsakis** [comment]
Sorry, it's been wild couple of weeks. I did some experimentation. I don't recall finding any example that could clarify what was making me nervous. 

In response to what @arielb1 wrote:

> BTW, it does not work ATM because of subtyping (the "is sized" relationship doesn't propagate through that) - should I change that?

I don't fully understand what you are proposed to change here. Can you elaborate? Ah, I guess I maybe see. The problem is that we wind up with a type variable for `!`, let's call it `?X` -- then we create a "subtype" variable `?Y` where `?Y <: ?X`, right? We know that `?X: Sized` but then we don't propagate that `?Y: Sized`, and hence we wind up the wrong coercion?

Interesting.

> I added the ty_var_root normalization to the closure code too.

Yes, I know, and it's probably good to do so. I tend to agree that we should use unification the way it was meant to be -- I am just basically trying to suss out possible places that we may become *more* stuck to implementation artifacts than we are today.

One thing that was making me nervous, for example, was that we might wind up with problems from poking at `pending_obligations` when we move to Chalk, since e.g. in the chalk style we wouldn't have the "confirmation" step that might replace a higher-level predicate like `T:  Foo` with the predicates from the impl.

I guess I need to schedule a slot to dig through and do more experimentation in that regard, but I'm definitely close to r+ here.

**Comment by arielb1** [comment]
>  We know that `?X: Sized` but then we don't propagate that `?Y: Sized`, and hence we wind up the wrong coercion?

Exactly that. We don't propagate that `?Y: Sized`, so we don't do our coercion.

**Comment by nikomatsakis** [comment]
@bors r+

OK. I did another read through the code and gave a bit of thought to alternative approaches -- I decided that I'm game to land this, but I would really like us to prioritize modeling the Rust type checker and its coercions (I *expect* that to be done in a chalk-like setting). It feels really important that we have a good handle on what the compiler knows in order to make its judgements at any time etc.

With regard to the subtyping rule, @arielb1, I'd rather take it slow. Let's open a FIXME perhaps and consider it as a follow-up?

I think in my ideal world we would try to model the type-checker in a chalk-like style, so that we can express these coercions in that world (and ideally try to address the problem of over-eager coercions as well, although I'm not sure how easy that will be). This would probably be best to do in the context of the Traits Working Group, I guess. Seems like maybe a good Rust All Hands discussion topic.

**Comment by arielb1** [comment]
```
[00:46:48] error: failed to download from `https://crates.io/api/v1/crates/openssl-src/111.1.0+1.1.1a/download`
[00:46:48] Caused by:
[00:46:48]   [6] Couldn't resolve host name (Could not resolve host: crates.io)
```

@bors retry

**Pull request merged**